### PR TITLE
Adding Orb command

### PR DESCRIPTION
**Pull request by ochaloup**
I'm proposing low level commands (online/offline) for orb subsystem.

### CODE REVIEW

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ * it could be
+ * <ul>
+ *   <li>having value - current value will be redefined</li>
+ *   <li>having no value (being null) - leaving current value where it is</li>
+ *   <li>undefined value - current value will be undefined</li>
+ * </ul>
+ */
+class Attribute<T> {
+    private T value;
+    private boolean isToUndefine = false;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java]: Should be easy to make the class `final` and these two fields `final` as well.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/AuthValues.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum AuthValues {
+    NONE("None"),
+    SERVER_AUTH("ServerAuth"),
+    CLIENT_AUTH("ClientAuth"),
+    MUTUAL_AUTH("MutualAuth");
+
+    public final String value;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/AuthValues.java]: Not `public`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+import org.wildfly.extras.creaper.core.online.operations.Address;
+import org.wildfly.extras.creaper.core.online.operations.Batch;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+
+/**
+ * Creaper command which allows settings of orb subsystem
+ *
+ * <ul>
+ *   <li>jacorb for EAP 6.x (<= WFLY 8)
+ *   <li>openjdk iiop for EAP 7.x (>= WFLY 9)
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java]: I'd prefer stating the upstream relationship first. If you want to add EAP relationship, that's fine, but _that_ belongs to the parens. Like:

```java
<li>JacORB for AS7 and WildFly <= 8 (EAP 6)</li>
<li>OpenJDK ORB for WildFly >= 9 (EAP 7)</li>
```

Personally, I don't feel the EAP bits are needed, but feel free to keep them.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    public boolean isUndefine() {
+        return isToUndefine;
+    }
+
+    /**
+     * Defines there is a value for this attribute.<br/>
+     * It does <b>not</b> consider if this attribute should or should not be undefined.
+     */
+    public boolean havingValue() {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java]: `hasValue`

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        return value != null;
+    }
+
+    /**
+     * Defines there is no value for this attribute.<br/>
+     * It does <b>not</b> consider if this attribute should or should not be undefined.
+     */
+    public boolean notHavingValue() {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java]: `hasNoValue` or `doesntHaveValue`

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+                .parameter("persistentServerId",
+                        orbType == OrbType.EAP_JACORB ? Attribute.noValue() : persistentServerId);
+        }
+
+        ctx.client.apply(transform.build());
+    }
+
+    static enum OrbType {
+        // jacorb in EAP 6.4.x series and lower
+        EAP_JACORB,
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java]: `AS7_JACORB`

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+        // jacorb in WFLY 8 and 9
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java]: So what is it? In `ChangeOrb` javadoc, you say that JacORB is <= WildFly 8 and OpenJDK ORB is WildFly >= 9.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+            return this;
+        }
+
+        public Builder persistentServerId(String persistentServerId) {
+            this.persistentServerId = Attribute.of(persistentServerId);
+            return this;
+        }
+
+
+
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java]: Why so many empty lines? :-)

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOffValues.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum OnOffValues {
+    ON(true, "on"),
+    OFF(false, "off");
+
+    public final Boolean openjdk;
+    public final String jacorb;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOffValues.java]: Not `public`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOffValues.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum OnOffValues {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOffValues.java]: `OnOff`

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/SecurityValues.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum SecurityValues {
+    IDENTITY("identity", "identity"),
+    CLIENT("client", "client"),
+    NONE("none", "off");
+
+    public final String openjdk;
+    public final String jacorb;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/SecurityValues.java]: Not `public`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/SupportedValues.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum SupportedValues {
+    NONE("none"),
+    SUPPORTED("supported"),
+    REQUIRED("required");
+
+    public final String value;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/SupportedValues.java]: Not `public`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/TransactionValues.java [hunk 13]
```diff
@@ -0,0 +1,15 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum TransactionValues {
+    ON("full", "on"),
+    OFF("none", "off"),
+    SPEC("spec", "spec");
+
+    public final String openjdk;
+    public final String jacorb;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/TransactionValues.java]: Not `public`.

**Comment by Ladicek** [comment]
I wonder why the test failed, but it shouldn't be caused by the changes in this PR. I'll look into it.

**Comment by Ladicek** [comment]
Could you please rebase on master? I've fixed the failing test.

**Comment by ochaloup** [comment]
I'm sorry for overhelming the ci server with wrong pushs. I changed the code at way you requested. Hopefully it's ok.

Regarding you remark on using `Attribute` class. I though that this is the right way how to create a command - having some low level that could be used for later addition of high level ones. I was absorbed with this idea that I spent quite significant effort to implement that way.

But with your note when looking back I have to say I'm now unsure if it was right path then. I can image that it would be nice to have all commands implemented in similar way but I understand it's first probably impossible because of time limitation and second I could be alone with this objective.

I fully understand that this way the command is different from other ones.
If you be happier if I split command to two `Change` and `Undefine` for not having the `Attribute` functionality involved I can do so.
On the other hand the public api does not deal with the trouble of some strange `Attribute` class being seen as you demanded.
What is your stand point on this?

**Comment by ochaloup** [comment]
ops, I'm reviewing old version of your notes. I need to need a bit more changes. Somehow overlooked that github is not in sync.

**Comment by Ladicek** [comment]
No, no need to change the architecture. Noone cares about configuring ORB anyway :-), so the minor difference doesn't really matter.

**Comment by ochaloup** [comment]
...now I'm not sure why tests failed. I run on my machine and it passes.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/AuthValues.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+        this.value = auth;
+    }
+
+    public String toString() {
+        return value;
+    }
+
+    public String getValue() {
+        return value;
+    }
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/AuthValues.java]: Why do you need this `public`? Either this method should be package-private or the `value` field should be package-private. I guess this also applies to other enums in this PR.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java [hunk 15]
```diff
@@ -0,0 +1,606 @@
...
+                        ? Attribute.of(addComponentViaInterceptor.get().getJacorb()) : addComponentViaInterceptor)
+                .parameter("persistentServerId",
+                        orbType == OrbType.AS7_JACORB ? Attribute.noValue() : persistentServerId);
+        }
+
+        ctx.client.apply(transform.build());
+    }
+
+    enum OrbType {
+        // jacorb in AS7 and WFLY 7
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java]: There's no WFLY 7.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    public Boolean getOpenJdk() {
+        return openjdk;
+    }
+
+    public String getJacorb() {
+        return jacorb;
+    }
+}
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java]: All `Boolean`s in this class could apparently be `boolean`s?

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/orb/ChangeOrbOnlineTest.java [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+import com.google.common.collect.ImmutableMap;
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+
+@Category(ManualTests.class)
+@RunWith(Arquillian.class)
+@RunAsClient
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/orb/ChangeOrbOnlineTest.java]: `@RunAsClient` is useless, please remove. The warnings from Arquillian that there's no running deployment are a bug, see [ARQ-2012](https://issues.jboss.org/browse/ARQ-2012).

**Comment by Ladicek** [comment]
Yea, I don't know why the test failed either. I'll try to figure it out.

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/orb/ChangeOrbOnlineTest.java (hunk 17 above)
Comment (ochaloup) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/orb/ChangeOrbOnlineTest.java]: Personally I like usage of `@RunAsClient` as denotes if the test was written with notion of running on client or not. But of course removing by your request :smile: 
Thanks for pointing to the issue.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java (hunk 15 above)
Comment (ochaloup) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java]: after so many years I should find out they were lying me? impossible!

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/AuthValues.java (hunk 14 above)
Comment (ochaloup) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/AuthValues.java]: ah, sure. I had to be unconscious when fixing the issue last time.

...don't worry master, I won't disappoint you again :]

**Comment by ochaloup** [comment]
ok, hopefully all requests fixed.

Unfortunately ci fails again. It passes locally for me. No much idea how to get some more info from travis ci...

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+    public static <T> Attribute<T> noValue() {
+        return new Attribute<T>(null);
+    }
+
+    public static <T> Attribute<T> undefine() {
+        return new Attribute<T>(null, true);
+    }
+
+    public static <O> Attribute<O> undefine(O object) {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java]: Hmm, I wonder what is this good for? The `object` is never used in this method. And the method seems to be unused as well.

If you wanted a type-safe method and the `<T> Attribute<T> undefine()` variant isn't enough (which might happen in Java pre 8), idiomatic solution would be `<T> Attribute<T> undefine(Class<? extends T> clazz)`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java [hunk 19]
```diff
@@ -0,0 +1,606 @@
...
+
+/**
+ * Creaper command which allows settings of orb subsystem
+ *
+ * <ul>
+ *   <li>JacORB for AS7 and WildFly <= 8</li>
+ *   <li>OpenJDK ORB for WildFly >= 9</li>
+ * </ul>
+ */
+public class ChangeOrb implements OfflineCommand, OnlineCommand {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java]: `final`?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java [hunk 20]
```diff
@@ -0,0 +1,606 @@
...
+
+        String subsystemName() {
+            return this == OrbType.WFLY_IIOP ? OPENJDK_IIOP_SUBSYSTEM_NAME : JACORB_SUBSYSTEM_NAME;
+        }
+    }
+
+
+    private static final class AttributeWriter {
+        private Address address;
+        private Batch batch;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java]: Both should be `final`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java [hunk 21]
```diff
@@ -0,0 +1,18 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum OnOff {
+    ON(true, "on"),
+    OFF(false, "off");
+
+    final boolean openjdk;
+    final String jacorb;
+
+    OnOff(Boolean openjdk, String jacorb) {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java]: `boolean`?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java (hunk 19 above)
Comment (ochaloup) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/ChangeOrb.java]: makes sence

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java (hunk 18 above)
Comment (ochaloup) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/Attribute.java]: it isn't used, I've removed it. It was left in code from pre-pull request dates where it was used in command builder.

my lack of knowledge. I was thinking about `Class<? extends T> clazz` but I got some trouble to do so which I can't remember. And I didn't know that is considered to be an idiomatic solution.
Btw. why this does not happen for Java 8 (for not stealing your time, some link would be enough :)?

btw. a longer time I'm wondering why `Arrays#toArray(T[] a)` method (https://docs.oracle.com/javase/7/docs/api/java/util/Collection.html#toArray(T[])) is not defined in similar way as you mentioned?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java (hunk 21 above)
Comment (ochaloup) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java]: agree, I was too rash

**Comment by ochaloup** [comment]
final adjustments

...final? LOL!

**Comment by Ladicek** [comment]
Tests seem to be failing on WildFly 8 and 9, please run `mvn clean verify -Pwildfly8` (resp. `-Pwildfly9`). Otherwise it's looking fine.

Oh, maybe you could fix the typo in the commit message -- it's not `codearc`, it's `CodeNarc` :-)

**Comment by ochaloup** [comment]
Commit message fixed.

Fixed the trouble with wfly8 and wfly9 - simplifier the options to have only jacorb or iiop. I expect now it should be working fine for all versions.

**Comment by Ladicek** [comment]
I tried to improve the `ShutdownTest` a bit, could you please rebase?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java [hunk 22]
```diff
@@ -0,0 +1,18 @@
+package org.wildfly.extras.creaper.commands.orb;
+
+public enum OnOff {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/orb/OnOff.java]: This doesn't seem to be exposed anywhere, so it shouldn't be `public`. Other enums seem to be used, so it's just this one.

**Comment by ochaloup** [comment]
Hiding OnOff and the rebase done.

**Comment by Ladicek** [comment]
Merged, finally :-) Thanks for the patience ;-)

**Comment by ochaloup** [comment]
My thanks to you.

 ...for patience ;) and for good advices!

**Pull request merged**